### PR TITLE
Feature/8 general project improvements

### PR DESCRIPTION
**Pull request by alexsrosa**

### CODE REVIEW

**Comment by FeatureSpitter** [comment]
Hello @alexsrosa ,

thank you for your submission!

I will start your review with the following question 😄 

What is the reason why you decided to not have a FK connecting the tables products and store_products?

imagem

**Comment by FeatureSpitter** [comment]
Given that after a bunch of 500 errors, the server starts responding again, why doesn't the attempts count reset?

imagem


Can you tell me what are the consequences of this counter not resetting?

**Comment by FeatureSpitter** [comment]
Any reason why the response payload for the /stores filter is not returning the `cluster` of each store?

imagem

**Comment by FeatureSpitter** [comment]
If you try to perform the following query:

```
POST http://localhost:8080/filters/STORE_NAME
{
  "filters": [
    {
      "id": "Season",
      "values": ["S18"]
    },
    {
      "id": "Cluster",
      "values": ["Europe"]
    }
  ]
}
```

You will notice it takes quite a few seconds to load. What techniques would you use in order to make this time significantly shorter?

Do you know what is the SQL query the ORM is performing to the database?

File: README.md [hunk 1]
```diff
@@ -1 +1,203 @@
 # Application SA. Service
+
```
Comment (FeatureSpitter) [inline README.md:2]: 👍 Nice README.

File: src/test/kotlin/pt/app/sa/service/controller/FiltersByClusterControllerTest.kt [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+/**
+ *
+ * @author <a href="mailto:alexsros@gmail.com">Alex Rosa</a>
+ * @since 28/09/2021 00:17
+ */
+@ActiveProfiles("test")
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@AutoConfigureMockMvc
+@Sql("/sql/import_test_data.sql")
```
Comment (FeatureSpitter) [inline src/test/kotlin/pt/app/sa/service/controller/FiltersByClusterControllerTest.kt:26]: 👍 Nice

File: src/main/kotlin/pt/app/sa/service/commons/ControlThreadsUtils.kt [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+ *
+ * @author <a href="mailto:alexsros@gmail.com">Alex Rosa</a>
+ * @since 03/10/2021 22:47
+ */
+class ControlThreadsUtils {
+
+    companion object {
+        fun returnWhenCloseAllTreadsOnList(listNameThreads: MutableList<Thread>) {
```
Comment (FeatureSpitter) [inline src/main/kotlin/pt/app/sa/service/commons/ControlThreadsUtils.kt:11]: A function like this wasn't necessary if you used kotlin coroutines.

File: src/main/kotlin/pt/app/sa/service/scheduler/LoadAndUpdateSourceSchedule.kt [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    val loadRegionsUseCase: LoadRegionsUseCase,
+    val loadStoresUseCase: LoadStoresUseCase,
+    val loadStoreProductsUseCase: LoadStoreProductsUseCase,
+    val loadProductsUseCase: LoadProductsUseCase,
+    @Value("\${scheduler.activeOnStart}") private val activeOnStart: Boolean
+) {
+
+    private val logger = LoggerFactory.getLogger(LoadAndUpdateSourceSchedule::class.java)
+
+    @PostConstruct
```
Comment (FeatureSpitter) [inline src/main/kotlin/pt/app/sa/service/scheduler/LoadAndUpdateSourceSchedule.kt:34]: This blocks the API from starting until the loadAndUpdateSourceSchedule execution ends.

How would you avoid this from happening?

**Comment by alexsrosa** [comment]
> Hello @alexsrosa ,
> 
> thank you for your submission!
> 
> I will start your review with the following question 😄
> 
> What is the reason why you decided to not have a FK connecting the tables products and store_products?
> 
> imagem

Hello @FeatureSpitter,

First of all thanks for your feedback :D

In this case I made a oneToMany relation in the StoreProductEntity class and it seems to have solved the need for now.

But I confess that I lacked understanding about what would be the correct key. Perhaps it should be the composite key of "season" and "model" from the products table versus "season" and "product" from the store_product table.

**Comment by alexsrosa** [comment]
> Can you tell me what are the consequences of this counter not resetting?

Good question. I thought about restarting the attempts as soon as the page was successful, but decided to keep the global attempts for each thead. That way I guarantee that we won't have an infinite loop if the external API doesn't come back. And, in return, we have a schedule that works every hour.

**Comment by FeatureSpitter** [comment]
> 
> 
> > Hello @alexsrosa ,
> > thank you for your submission!
> > I will start your review with the following question 😄
> > What is the reason why you decided to not have a FK connecting the tables products and store_products?
> > imagem
> 
> Hello @FeatureSpitter,
> 
> First of all thanks for your feedback :D
> 
> In this case I made a oneToMany relation in the StoreProductEntity class and it seems to have solved the need for now.
> 
> But I confess that I lacked understanding about what would be the correct key. Perhaps it should be the composite key of "season" and "model" from the products table versus "season" and "product" from the store_product table.

Ah, I forgot to mention that as well: Although surrogate ID keys are the best practice here (also product -> product_id instead), you didn't add a UNIQUE constraint on the tables based on their natural characteristics. For example in this table the UNIQUE constraint would be `(product_id, store_id, season)`, to avoid duplicate relations (and it would also allow you to make use of the amazingly efficient on-conflict-do-update instruction from Postgresql). 😄

**Comment by FeatureSpitter** [comment]
> 
> 
> > Can you tell me what are the consequences of this counter not resetting?
> 
> Good question. I thought about restarting the attempts as soon as the page was successful, but decided to keep the global attempts for each thead. That way I guarantee that we won't have an infinite loop if the external API doesn't come back. And, in return, we have a schedule that works every hour.

You would not end up with an infinite loop, given that the counter would reset only if the request had success.

For eg.:

```
Error #1 - Counter = 1/1200
Error #2 - Counter = 2/1200
Success - Counter = 0/1200 (resets)
Success - Counter = 0/1200 (keeps at 0)
Error #3 - Counter = 1/1200
Error #4 - Counter = 2/1200
Error #5 - Counter = 3/1200
etc...
```

The worst case scenario is if every single request fails 1199 times and succeeds at last, but this is not infinite. Although, like you said, it could take so long it could collide with the next scheduled execution... But for this I would just lower 1200 to a smaller number (eg. 5), or other "magic number". Why 1200 😆 ?

**Comment by alexsrosa** [comment]
> Any reason why the response payload for the /stores filter is not returning the `cluster` of each store?
> 
> imagem

It was my mistake, I will send a new commit with the adjustment.

File: src/main/kotlin/pt/app/sa/service/commons/ControlThreadsUtils.kt (hunk 3 above)
Comment (alexsrosa) [reply src/main/kotlin/pt/app/sa/service/commons/ControlThreadsUtils.kt:11]: Yes, I ended up not using this feature for lack of practice with Kotlin, but I read the documentation at the time.

File: src/main/kotlin/pt/app/sa/service/scheduler/LoadAndUpdateSourceSchedule.kt (hunk 4 above)
Comment (alexsrosa) [reply src/main/kotlin/pt/app/sa/service/scheduler/LoadAndUpdateSourceSchedule.kt:34]: True, I ended up not getting it. My suggestion is to put in a new Thead. In the real world this execution might not be necessary at start, as only cron would be more correct.

**Comment by alexsrosa** [comment]
> > > Hello @alexsrosa ,
> > > thank you for your submission!
> > > I will start your review with the following question 😄
> > > What is the reason why you decided to not have a FK connecting the tables products and store_products?
> > > imagem
> > 
> > 
> > Hello @FeatureSpitter,
> > First of all thanks for your feedback :D
> > In this case I made a oneToMany relation in the StoreProductEntity class and it seems to have solved the need for now.
> > But I confess that I lacked understanding about what would be the correct key. Perhaps it should be the composite key of "season" and "model" from the products table versus "season" and "product" from the store_product table.
> 
> Ah, I forgot to mention that as well: Although surrogate ID keys are the best practice here (also product -> product_id instead), you didn't add a UNIQUE constraint on the tables based on their natural characteristics. For example in this table the UNIQUE constraint would be `(product_id, store_id, season)`, to avoid duplicate relations (and it would also allow you to make use of the amazingly efficient on-conflict-do-update instruction from Postgresql). 😄

I agree, it would be better as you suggested. But I also created a unique with product (but the name), season and store_id. To update, I created a hash_id, which made it much easier to search and validate if there was a need for an update.

**Comment by alexsrosa** [comment]
> > > Can you tell me what are the consequences of this counter not resetting?
> > 
> > 
> > Good question. I thought about restarting the attempts as soon as the page was successful, but decided to keep the global attempts for each thead. That way I guarantee that we won't have an infinite loop if the external API doesn't come back. And, in return, we have a schedule that works every hour.
> 
> You would not end up with an infinite loop, given that the counter would reset only if the request had success.
> 
> For eg.:
> 
> ```
> Error #1 - Counter = 1/1200
> Error #2 - Counter = 2/1200
> Success - Counter = 0/1200 (resets)
> Success - Counter = 0/1200 (keeps at 0)
> Error #3 - Counter = 1/1200
> Error #4 - Counter = 2/1200
> Error #5 - Counter = 3/1200
> etc...
> ```
> 
> The worst case scenario is if every single request fails 1199 times and succeeds at last, but this is not infinite. Although, like you said, it could take so long it could collide with the next scheduled execution... But for this I would just lower 1200 to a smaller number (eg. 5), or other "magic number". Why 1200 😆 ?

True, so, in this case, the best approach would be to reset the counter and leave it with a low value per page.
The symbolic value of "1200" ended up remaining, but at the beginning of development I was getting a lot of errors and not working with theads, so in many cases I reached this value and finally not loading the entire database. But now we can lessen that.

**Comment by alexsrosa** [comment]
> If you try to perform the following query:
> 
> ```
> POST http://localhost:8080/filters/STORE_NAME
> {
>   "filters": [
>     {
>       "id": "Season",
>       "values": ["S18"]
>     },
>     {
>       "id": "Cluster",
>       "values": ["Europe"]
>     }
>   ]
> }
> ```
> 
> You will notice it takes quite a few seconds to load. What techniques would you use in order to make this time significantly shorter?
> 
> Do you know what is the SQL query the ORM is performing to the database?

It really is a problem, well caught. This happens because of the 'DISTINCT' used in the query. This causes full access to related tables. A suggestion would be to create a subquery in the join with the store_product table already with the distinct and also use group by.

**Query created by ORM:** 8s :(((

```
select
    DISTINCT storeentit0_.id as id1_4_,
    storeentit0_.created_at as created_2_4_,
    storeentit0_.updated_at as updated_3_4_,
    storeentit0_.name as name4_4_,
    storeentit0_.name_alias as name_ali5_4_,
    storeentit0_.region_id as region_i7_4_,
    storeentit0_.theme as theme6_4_
from stores storeentit0_
 inner join store_products storeprodu1_ on storeentit0_.id=storeprodu1_.store_id
 inner join regions regionenti2_ on storeentit0_.region_id=regionenti2_.id
 inner join clusters clusterent3_ on regionenti2_.clusters_id=clusterent3_.id
where
    ( storeprodu1_.season in ('S18' ) )
  and ( clusterent3_.name in ( 'Europe' ))
order by
    storeentit0_.name_alias asc limit 100;
```

**Suggested query:** 315ms :D

```
select
    storeentit0_.id as id1_4_,
    storeentit0_.created_at as created_2_4_,
    storeentit0_.updated_at as updated_3_4_,
    storeentit0_.name as name4_4_,
    storeentit0_.name_alias as name_ali5_4_,
    storeentit0_.region_id as region_i7_4_,
    storeentit0_.theme as theme6_4_
from stores storeentit0_
         inner join (select distinct store_id, season from store_products) storeprodu1_ on storeentit0_.id=storeprodu1_.store_id
         inner join regions regionenti2_ on storeentit0_.region_id=regionenti2_.id
         inner join clusters clusterent3_ on regionenti2_.clusters_id=clusterent3_.id
where
    ( storeprodu1_.season in ('S18' ) )
  and ( clusterent3_.name in ( 'Europe' ))
group by
    storeentit0_.id ,
    storeentit0_.name ,
    storeentit0_.name_alias ,
    storeentit0_.region_id ,
    storeentit0_.theme
order by
    storeentit0_.name_alias asc limit 100;
```

**Comment by alexsrosa** [comment]
Sorry, for some reason I was wrong and closed the PR :(

**Comment by FeatureSpitter** [comment]
> 
> 
> > If you try to perform the following query:
> > ```
> > POST http://localhost:8080/filters/STORE_NAME
> > {
> >   "filters": [
> >     {
> >       "id": "Season",
> >       "values": ["S18"]
> >     },
> >     {
> >       "id": "Cluster",
> >       "values": ["Europe"]
> >     }
> >   ]
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You will notice it takes quite a few seconds to load. What techniques would you use in order to make this time significantly shorter?
> > Do you know what is the SQL query the ORM is performing to the database?
> 
> It really is a problem, well caught. This happens because of the 'DISTINCT' used in the query. This causes full access to related tables. A suggestion would be to create a subquery in the join with the store_product table already with the distinct and also use group by.
> 
> **Query created by ORM:** 8s :(((
> 
> ```
> select
>     DISTINCT storeentit0_.id as id1_4_,
>     storeentit0_.created_at as created_2_4_,
>     storeentit0_.updated_at as updated_3_4_,
>     storeentit0_.name as name4_4_,
>     storeentit0_.name_alias as name_ali5_4_,
>     storeentit0_.region_id as region_i7_4_,
>     storeentit0_.theme as theme6_4_
> from stores storeentit0_
>  inner join store_products storeprodu1_ on storeentit0_.id=storeprodu1_.store_id
>  inner join regions regionenti2_ on storeentit0_.region_id=regionenti2_.id
>  inner join clusters clusterent3_ on regionenti2_.clusters_id=clusterent3_.id
> where
>     ( storeprodu1_.season in ('S18' ) )
>   and ( clusterent3_.name in ( 'Europe' ))
> order by
>     storeentit0_.name_alias asc limit 100;
> ```
> 
> **Suggested query:** 315ms :D
> 
> ```
> select
>     storeentit0_.id as id1_4_,
>     storeentit0_.created_at as created_2_4_,
>     storeentit0_.updated_at as updated_3_4_,
>     storeentit0_.name as name4_4_,
>     storeentit0_.name_alias as name_ali5_4_,
>     storeentit0_.region_id as region_i7_4_,
>     storeentit0_.theme as theme6_4_
> from stores storeentit0_
>          inner join (select distinct store_id, season from store_products) storeprodu1_ on storeentit0_.id=storeprodu1_.store_id
>          inner join regions regionenti2_ on storeentit0_.region_id=regionenti2_.id
>          inner join clusters clusterent3_ on regionenti2_.clusters_id=clusterent3_.id
> where
>     ( storeprodu1_.season in ('S18' ) )
>   and ( clusterent3_.name in ( 'Europe' ))
> group by
>     storeentit0_.id ,
>     storeentit0_.name ,
>     storeentit0_.name_alias ,
>     storeentit0_.region_id ,
>     storeentit0_.theme
> order by
>     storeentit0_.name_alias asc limit 100;
> ```

👍 Perfect!

This is one of the major reasons why we started using Ktorm for these kind of projects instead of Hibernate. The other major reason was the Criteria API, which is bad when compared to Ktorm's. Still, we are not 100% happy with Ktorm as it lacks full support when linking the code with existing schemas (or generating the schema from the code).

**Pull request merged**